executor/mem_reader.go

```diff
@@ -525,3 +545,188 @@ func (m *memIndexLookUpReader) getMemRows() ([][]types.Datum, error) {
 
 	return memTblReader.getMemRows()
 }
+
+type memIndexMergeReader struct {
+	ctx              sessionctx.Context
+	index            *model.IndexInfo
+	columns          []*model.ColumnInfo
+	table            table.Table
+	desc             bool
+	conditions       []expression.Expression
+	retFieldTypes    []*types.FieldType
+	indexMergeReader *IndexMergeReaderExecutor
+
+	memIndexReaders []*memIndexReader
+	memTableReaders []*memTableReader
+
+	// partition mode
+	partitionMode     bool                  // if it is accessing a partition table
+	partitionTables   []table.PhysicalTable // partition tables to access
+	partitionKVRanges [][][]kv.KeyRange     // kv ranges for these partition tables
+}
+
+func buildMemIndexMergeReader(us *UnionScanExec, indexMergeReader *IndexMergeReaderExecutor) *memIndexMergeReader {
+	indexCount := len(indexMergeReader.indexes)
+	indexReaders := make([]*memIndexReader, 0, indexCount)
+	tableReaders := make([]*memTableReader, 0, indexCount)
+	for i := 0; i < indexCount; i++ {
+		if indexMergeReader.indexes[i] == nil {
+			colIDs, pkColIDs, rd := getColIDAndPkColIDs(indexMergeReader.table, indexMergeReader.columns)
+			tableReaders = append(tableReaders, &memTableReader{
+				ctx:           us.ctx,
+				table:         indexMergeReader.table.Meta(),
+				columns:       indexMergeReader.columns,
+				kvRanges:      nil,
+				conditions:    us.conditions,
+				addedRows:     make([][]types.Datum, 0),
+				retFieldTypes: retTypes(us),
+				colIDs:        colIDs,
+				pkColIDs:      pkColIDs,
+				buffer: allocBuf{
+					handleBytes: make([]byte, 0, 16),
+					rd:          rd,
+				},
+				handleCols: indexMergeReader.handleCols,
+			})
+			indexReaders = append(indexReaders, nil)
+		} else {
+			outputOffset := []int{len(indexMergeReader.indexes[i].Columns)}
+			indexReaders = append(indexReaders, &memIndexReader{
+				ctx:             us.ctx,
+				index:           indexMergeReader.indexes[i],
+				table:           indexMergeReader.table.Meta(),
+				kvRanges:        nil,
+				desc:            indexMergeReader.descs[i],
+				retFieldTypes:   retTypes(us),
+				outputOffset:    outputOffset,
+				belowHandleCols: us.belowHandleCols,
+			})
+			tableReaders = append(tableReaders, nil)
+		}
+	}
```

Comment:
Could we extract an interface to merge this if branch?

Reply:
I'm not sure if you mean the following:

    for i := 0; i < indexCount; i++ {
        // if partialPlans[i] is IndexScan, this is an invalid tableReader.
        tableReaders = append(tableReaders, &memTableReader{...});
        // if partialPlans[i] is not IndexScan, this is a invalid indexReader.
        indexReaders = append(indexReaders, &memIndexReader{...}); 
    }

However, this may lead to the need to construct additional tableReader or indexReader?

Reply:
Oh, i mean the one like this
```
for i := 0; i < partialPathCount; i ++ {
	partialReaders = append(partialReaders, index/table reader)
}
```

Reply:
I extract an interface `memReader`, but the if branch still exists. But we only need one slice to store memTableReader/memIndexReader.